experiment/secret-sync-controller/main.go

```diff
@@ -0,0 +1,219 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	// apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog"
+	"k8s.io/test-infra/experiment/secret-sync-controller/client"
+	"os"
+	"time"
+)
+
+type options struct {
+	configPath   string
+	testSetup    string
+	runOnce      bool
+	resyncPeriod int64
+}
+
+func (o *options) Validate() error {
+	if o.configPath == "" {
+		return errors.New("required flag --config-path was unset")
+	}
+	return nil
+}
+
+func gatherOptions() options {
+	o := options{}
+	flag.StringVar(&o.configPath, "config-path", "", "Path to config.yaml.")
+	flag.StringVar(&o.testSetup, "test-setup", "", "Path to test-setup.yaml.")
+	flag.BoolVar(&o.runOnce, "run-once", false, "Sync once instead of continuous loop.")
+	flag.Int64Var(&o.resyncPeriod, "period", 1000, "Resync period in milliseconds.")
+	flag.Parse()
+	return o
+}
+
+type SecretSyncController struct {
+	Client       client.Interface
+	Config       *SecretSyncConfig
+	RunOnce      bool
+	ResyncPeriod time.Duration
+	// TODO: in-struct looger?
+}
+
+func main() {
+	o := gatherOptions()
+
+	// prepare clients
+	k8sClientset, err := client.NewK8sClientset()
+	if err != nil {
+		klog.Errorf("New kubernetes client failed: %s", err)
+	}
+	secretManagerClient, err := client.NewSecretManagerClient(context.Background())
+	if err != nil {
+		klog.Errorf("New Secret Manager client failed: %s", err)
+	}
+	clientInterface := &client.Client{
+		K8sClientset:        *k8sClientset,
+		SecretManagerClient: *secretManagerClient,
+	}
+
+	// prepare config
+	secretSyncConfig, err := LoadConfig(o.configPath)
```

Comment:
Once you've loaded the config, you're going to want to validate it to make sure there are no invalid specs, eg:
- a destination appears twice
- there are no empty field names

You might want to create a skeleton/bare bones `SecretSyncConfig.Validate()` function to fill out later

Reply:
Done